File: packages/server-python/app/routers/milito.py
"""Milito game routes."""

import random
from datetime import datetime
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies import get_current_user
from app.models.game import Game, GamePlayer, GameRoom, GameState
from app.models.user import User
from app.schemas.milito import DiscardRequest, SelectCardRequest, SelectColumnRequest

router = APIRouter(prefix="/milito", tags=["milito"])

# Game constants
MILITO_SELECT_CARD = "SELECT_CARD"
MILITO_SELECT_COLUMN = "SELECT_COLUMN"
MILITO_DISCARD = "DISCARD"

FACTIONS = ["ancient_british", "alexandrian_macedonian"]


def create_initial_table() -> dict:
    """Create an empty battlefield table."""
    return {
        "enemy_row_2": [None, None, None, None, None],
        "enemy_row_1": [None, None, None, None, None],
        "territory_row": [0, 0, 0, 0, 0],
        "player_row_1": [None, None, None, None, None],
        "player_row_2": [None, None, None, None, None],
    }


def create_initial_hand() -> list[int]:
    """Create initial hand with 5 random cards (0-4 are unit type indices)."""
    return [random.randint(0, 4) for _ in range(5)]


def create_player_state(player_id: str, faction: str) -> dict:
    """Create initial state for a player."""
    return {
        "oderId": player_id,
        "faction": faction,
        "table": create_initial_table(),
        "hand": create_initial_hand(),
        "phase": MILITO_SELECT_CARD,
        "cardsToDiscard": 0,
        "selectedCard": None,
        "selectedColumn": None,
        "discardedCards": [],
        "score": 0,
    }


@router.get("/{room_id}/state")
def get_milito_state(
    room_id: str,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    """Get the current Milito game state."""
    # Verify player is in the room
    player = (
        db.query(GamePlayer)
        .filter(GamePlayer.room_id == room_id, GamePlayer.user_id == user.id)
        .first()
    )
    if not player:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You are not in this room",
        )

    # Get game state
    state_row = db.query(GameState).filter(GameState.room_id == room_id).first()
    if not state_row:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Game not started yet",
        )

    # Get room
    room = db.query(GameRoom).filter(GameRoom.id == room_id).first()

    # Get all players
    players = (
        db.query(GamePlayer)
        .filter(GamePlayer.room_id == room_id)
        .order_by(GamePlayer.position)
        .all()
    )

    state = state_row.state
    my_state = state["players"].get(user.id)

    return {
        "room": {
            "id": room.id,
            "gameId": room.game_id,
            "name": room.name,
            "hostId": room.host_id,
            "status": room.status,
            "maxPlayers": room.max_players,
        },
        "gameState": state,
        "myState": my_state,
        "players": [
            {"id": p.id, "userId": p.user_id, "position": p.position, "score": p.score}
            for p in players
        ],
        "isMyTurn": state["currentPlayerId"] == user.id,
    }


@router.post("/{room_id}/initialize")
def initialize_milito(
    room_id: str,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    """Initialize the Milito game (host only)."""
    room = db.query(GameRoom).filter(GameRoom.id == room_id).first()
    if not room:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Room not found",
        )

    # Verify this is a Milito room
    game = db.query(Game).filter(Game.id == room.game_id).first()
    if not game or game.id != "milito":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="This room is not a Milito game",
        )

    if room.host_id != user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only host can initialize the game",
        )

    # Get players
    players = db.query(GamePlayer).filter(GamePlayer.room_id == room_id).all()

    if len(players) != 2:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Milito requires exactly 2 players",
        )

    # Create player states with assigned factions
    player_states = {}
    for idx, player in enumerate(players):
        player_states[player.user_id] = create_player_state(player.user_id, FACTIONS[idx])

    # Create initial game state
    initial_state = {
        "players": player_states,
        "currentPlayerId": players[0].user_id,
        "turnNumber": 1,
        "status": "playing",
    }

    # Upsert game state
    state_row = db.query(GameState).filter(GameState.room_id == room_id).first()
    if state_row:
        state_row.state = initial_state
        state_row.current_turn = 1
        state_row.updated_at = datetime.utcnow()
    else:
        state_row = GameState(room_id=room_id, current_turn=1)
        state_row.state = initial_state
        db.add(state_row)

    # Update room status
    room.status = "playing"

    db.commit()
    return {"success": True}


@router.post("/{room_id}/select-card")
def select_card(
    room_id: str,
    body: SelectCardRequest,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    """Select a card from hand."""
    state_row = db.query(GameState).filter(GameState.room_id == room_id).first()
    if not state_row:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Game not started",
        )

    state = state_row.state

    # Check if it's this player's turn
    if state["currentPlayerId"] != user.id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Not your turn",
        )

    player_state = state["players"][user.id]
    if player_state["phase"] != MILITO_SELECT_CARD:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Not in card selection phase",
        )

    if body.cardIndex < 0 or body.cardIndex >= len(player_state["hand"]):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid card index",
        )

    # Update player state
    player_state["selectedCard"] = body.cardIndex
    player_state["phase"] = MILITO_SELECT_COLUMN

    # Save state
    state_row.state = state
    state_row.updated_at = datetime.utcnow()
    db.commit()

    return {"success": True, "state": state}


@router.post("/{room_id}/select-column")
def select_column(
    room_id: str,
    body: SelectColumnRequest,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    """Select a column to place the card."""
    state_row = db.query(GameState).filter(GameState.room_id == room_id).first()
    if not state_row:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Game not started",
        )

    state = state_row.state

    # Check if it's this player's turn
    if state["currentPlayerId"] != user.id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Not your turn",
        )

    player_state = state["players"][user.id]
    if player_state["phase"] != MILITO_SELECT_COLUMN:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Not in column selection phase",
        )

    if body.columnIndex < 0 or body.columnIndex >= 5:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid column index",
        )

    # Update player state
    player_state["selectedColumn"] = body.columnIndex
    player_state["cardsToDiscard"] = 1
    player_state["phase"] = MILITO_DISCARD

    # Save state
    state_row.state = state
    state_row.updated_at = datetime.utcnow()
    db.commit()

    return {"success": True, "state": state}


@router.post("/{room_id}/discard")
def discard_card(
    room_id: str,
    body: DiscardRequest,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    """Discard a card and complete the turn."""
    state_row = db.query(GameState).filter(GameState.room_id == room_id).first()
    if not state_row:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Game not started",
        )

    state = state_row.state

    # Check if it's this player's turn
    if state["currentPlayerId"] != user.id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Not your turn",
        )

    player_state = state["players"][user.id]
    if player_state["phase"] != MILITO_DISCARD:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Not in discard phase",
        )

    if body.cardIndex < 0 or body.cardIndex >= len(player_state["hand"]):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid card index",
        )

    # Can't discard the selected card
    if body.cardIndex == player_state["selectedCard"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Cannot discard the card you are playing",
        )

    player_state["discardedCards"].append(body.cardIndex)
    player_state["cardsToDiscard"] -= 1

    if player_state["cardsToDiscard"] <= 0:
        # Execute the turn: place card on board
        card_value = player_state["hand"][player_state["selectedCard"]]
        column = player_state["selectedColumn"]

        # Place card in player_row_1 if empty, otherwise player_row_2
        if player_state["table"]["player_row_1"][column] is None:
            player_state["table"]["player_row_1"][column] = card_value
        elif player_state["table"]["player_row_2"][column] is None:
            player_state["table"]["player_row_2"][column] = card_value

        # Remove discarded and played cards from hand
        indices_to_remove = set(player_state["discardedCards"])
        indices_to_remove.add(player_state["selectedCard"])
        new_hand = [
            card
            for idx, card in enumerate(player_state["hand"])
            if idx not in indices_to_remove
        ]

        # Draw cards to get back to 5
        while len(new_hand) < 5:
            new_hand.append(random.randint(0, 4))

        player_state["hand"] = new_hand
        player_state["phase"] = MILITO_SELECT_CARD
        player_state["selectedCard"] = None
        player_state["selectedColumn"] = None
        player_state["discardedCards"] = []

        # Switch to other player
        player_ids = list(state["players"].keys())
        current_index = player_ids.index(user.id)
        state["currentPlayerId"] = player_ids[(current_index + 1) % len(player_ids)]
        state["turnNumber"] += 1

    # Save state
    state_row.state = state
    state_row.current_turn = state["turnNumber"]
    state_row.updated_at = datetime.utcnow()
    db.commit()

    return {"success": True, "state": state}

File: src/server/routes/milito.ts
import { ApiRouter, t } from '../api-router'
import { db } from '../db/index.js'
import { gameRooms, gamePlayers, gameState, games } from '../db/schema.js'
import { eq, and } from 'drizzle-orm'
import { authMiddleware } from '../middleware/auth.js'
import type {
  MilitoGameState,
  MilitoPlayerState,
  MilitoPlayerTable,
  MilitoFaction,
} from '#/shared'

const MILITO_SELECT_CARD = 'SELECT_CARD'
const MILITO_SELECT_COLUMN = 'SELECT_COLUMN'
const MILITO_DISCARD = 'DISCARD'

function createInitialTable(): MilitoPlayerTable {
  return {
    enemy_row_2: [null, null, null, null, null],
    enemy_row_1: [null, null, null, null, null],
    territory_row: [0, 0, 0, 0, 0],
    player_row_1: [null, null, null, null, null],
    player_row_2: [null, null, null, null, null],
  }
}

function createInitialHand(): number[] {
  // Draw 5 random cards (0-4 are unit type indices)
  return Array.from({ length: 5 }, () => Math.floor(Math.random() * 5))
}

function createPlayerState(playerId: string, faction: MilitoFaction): MilitoPlayerState {
  return {
    oderId: playerId,
    faction,
    table: createInitialTable(),
    hand: createInitialHand(),
    phase: MILITO_SELECT_CARD,
    cardsToDiscard: 0,
    selectedCard: undefined,
    selectedColumn: undefined,
    discardedCards: [],
    score: 0,
  }
}

export const militoRoutes = new ApiRouter({ prefix: '/api/milito' })
  .use(authMiddleware)
  .get(
    '/:roomId/state',
    async ({ params, user, set }) => {
      if (!user) {
        set.status = 401
        return { error: 'Unauthorized', message: 'You must be logged in' }
      }

      const player = await db
        .select()
        .from(gamePlayers)
        .where(and(eq(gamePlayers.roomId, params.roomId), eq(gamePlayers.userId, user.id)))
        .limit(1)

      if (!player[0]) {
        set.status = 403
        return { error: 'Forbidden', message: 'You are not in this room' }
      }

      const stateRow = await db
        .select()
        .from(gameState)
        .where(eq(gameState.roomId, params.roomId))
        .limit(1)

      if (!stateRow[0]) {
        set.status = 404
        return { error: 'Not found', message: 'Game not started yet' }
      }

      const room = await db
        .select()
        .from(gameRooms)
        .where(eq(gameRooms.id, params.roomId))
        .limit(1)

      const players = await db
        .select()
        .from(gamePlayers)
        .where(eq(gamePlayers.roomId, params.roomId))
        .orderBy(gamePlayers.position)

      const state = stateRow[0].state as MilitoGameState
      const myPlayerState = state.players[user.id]

      return {
        room: room[0],
        gameState: state,
        myState: myPlayerState,
        players,
        isMyTurn: state.currentPlayerId === user.id,
      }
    },
    {
      params: t.Object({
        roomId: t.String(),
      }),
    },
  )
  .post(
    '/:roomId/initialize',
    async ({ params, user, set }) => {
      if (!user) {
        set.status = 401
        return { error: 'Unauthorized', message: 'You must be logged in' }
      }

      const room = await db
        .select()
        .from(gameRooms)
        .where(eq(gameRooms.id, params.roomId))
        .limit(1)

      if (!room[0]) {
        set.status = 404
        return { error: 'Not found', message: 'Room not found' }
      }

      // Verify this is a Milito room
      const game = await db.select().from(games).where(eq(games.id, room[0].gameId)).limit(1)
      if (!game[0] || game[0].id !== 'milito') {
        set.status = 400
        return { error: 'Bad request', message: 'This room is not a Milito game' }
      }

      if (room[0].hostId !== user.id) {
        set.status = 403
        return { error: 'Forbidden', message: 'Only host can initialize the game' }
      }

      const players = await db.select().from(gamePlayers).where(eq(gamePlayers.roomId, params.roomId))

      if (players.length !== 2) {
        set.status = 400
        return { error: 'Bad request', message: 'Milito requires exactly 2 players' }
      }

      const factions: MilitoFaction[] = ['ancient_british', 'alexandrian_macedonian']
      const playerStates: Record<string, MilitoPlayerState> = {}

      players.forEach((player, idx) => {
        playerStates[player.userId] = createPlayerState(player.userId, factions[idx])
      })

      const initialState: MilitoGameState = {
        players: playerStates,
        currentPlayerId: players[0].userId,
        turnNumber: 1,
        status: 'playing',
      }

      await db
        .insert(gameState)
        .values({
          roomId: params.roomId,
          state: initialState,
          currentTurn: 1,
        })
        .onConflictDoUpdate({
          target: gameState.roomId,
          set: {
            state: initialState,
            currentTurn: 1,
            updatedAt: new Date(),
          },
        })

      await db
        .update(gameRooms)
        .set({ status: 'playing' })
        .where(eq(gameRooms.id, params.roomId))

      return { success: true }
    },
    {
      params: t.Object({
        roomId: t.String(),
      }),
    },
  )
  .post(
    '/:roomId/select-card',
    async ({ params, body, user, set }) => {
      if (!user) {
        set.status = 401
        return { error: 'Unauthorized', message: 'You must be logged in' }
      }

      const stateRow = await db
        .select()
        .from(gameState)
        .where(eq(gameState.roomId, params.roomId))
        .limit(1)

      if (!stateRow[0]) {
        set.status = 404
        return { error: 'Not found', message: 'Game not started' }
      }

      const state = stateRow[0].state as MilitoGameState

      if (state.currentPlayerId !== user.id) {
        set.status = 400
        return { error: 'Bad request', message: 'Not your turn' }
      }

      const playerState = state.players[user.id]
      if (playerState.phase !== MILITO_SELECT_CARD) {
        set.status = 400
        return { error: 'Bad request', message: 'Not in card selection phase' }
      }

      if (body.cardIndex < 0 || body.cardIndex >= playerState.hand.length) {
        set.status = 400
        return { error: 'Bad request', message: 'Invalid card index' }
      }

      playerState.selectedCard = body.cardIndex
      playerState.phase = MILITO_SELECT_COLUMN

      await db
        .update(gameState)
        .set({ state, updatedAt: new Date() })
        .where(eq(gameState.roomId, params.roomId))

      return { success: true, state }
    },
    {
      params: t.Object({ roomId: t.String() }),
      body: t.Object({ cardIndex: t.Number() }),
    },
  )
  .post(
    '/:roomId/select-column',
    async ({ params, body, user, set }) => {
      if (!user) {
        set.status = 401
        return { error: 'Unauthorized', message: 'You must be logged in' }
      }

      const stateRow = await db
        .select()
        .from(gameState)
        .where(eq(gameState.roomId, params.roomId))
        .limit(1)

      if (!stateRow[0]) {
        set.status = 404
        return { error: 'Not found', message: 'Game not started' }
      }

      const state = stateRow[0].state as MilitoGameState

      if (state.currentPlayerId !== user.id) {
        set.status = 400
        return { error: 'Bad request', message: 'Not your turn' }
      }

      const playerState = state.players[user.id]
      if (playerState.phase !== MILITO_SELECT_COLUMN) {
        set.status = 400
        return { error: 'Bad request', message: 'Not in column selection phase' }
      }

      if (body.columnIndex < 0 || body.columnIndex >= 5) {
        set.status = 400
        return { error: 'Bad request', message: 'Invalid column index' }
      }

      playerState.selectedColumn = body.columnIndex
      playerState.cardsToDiscard = 1
      playerState.phase = MILITO_DISCARD

      await db
        .update(gameState)
        .set({ state, updatedAt: new Date() })
        .where(eq(gameState.roomId, params.roomId))

      return { success: true, state }
    },
    {
      params: t.Object({ roomId: t.String() }),
      body: t.Object({ columnIndex: t.Number() }),
    },
  )
  .post(
    '/:roomId/discard',
    async ({ params, body, user, set }) => {
      if (!user) {
        set.status = 401
        return { error: 'Unauthorized', message: 'You must be logged in' }
      }

      const stateRow = await db
        .select()
        .from(gameState)
        .where(eq(gameState.roomId, params.roomId))
        .limit(1)

      if (!stateRow[0]) {
        set.status = 404
        return { error: 'Not found', message: 'Game not started' }
      }

      const state = stateRow[0].state as MilitoGameState

      if (state.currentPlayerId !== user.id) {
        set.status = 400
        return { error: 'Bad request', message: 'Not your turn' }
      }

      const playerState = state.players[user.id]
      if (playerState.phase !== MILITO_DISCARD) {
        set.status = 400
        return { error: 'Bad request', message: 'Not in discard phase' }
      }

      if (body.cardIndex < 0 || body.cardIndex >= playerState.hand.length) {
        set.status = 400
        return { error: 'Bad request', message: 'Invalid card index' }
      }

      // Can't discard the selected card
      if (body.cardIndex === playerState.selectedCard) {
        set.status = 400
        return { error: 'Bad request', message: 'Cannot discard the card you are playing' }
      }

      playerState.discardedCards.push(body.cardIndex)
      playerState.cardsToDiscard--

      if (playerState.cardsToDiscard <= 0) {
        // Execute the turn: place card on board
        const cardValue = playerState.hand[playerState.selectedCard!]
        const column = playerState.selectedColumn!

        // Place card in player_row_1 if empty, otherwise player_row_2
        if (playerState.table.player_row_1[column] === null) {
          playerState.table.player_row_1[column] = cardValue
        } else if (playerState.table.player_row_2[column] === null) {
          playerState.table.player_row_2[column] = cardValue
        }

        // Remove discarded and played cards from hand
        const indicesToRemove = new Set([...playerState.discardedCards, playerState.selectedCard!])
        const newHand = playerState.hand.filter((_, idx) => !indicesToRemove.has(idx))

        // Draw cards to get back to 5
        while (newHand.length < 5) {
          newHand.push(Math.floor(Math.random() * 5))
        }

        playerState.hand = newHand
        playerState.phase = MILITO_SELECT_CARD
        playerState.selectedCard = undefined
        playerState.selectedColumn = undefined
        playerState.discardedCards = []

        // Switch to other player
        const playerIds = Object.keys(state.players)
        const currentIndex = playerIds.indexOf(user.id)
        state.currentPlayerId = playerIds[(currentIndex + 1) % playerIds.length]
        state.turnNumber++
      }

      await db
        .update(gameState)
        .set({ state, currentTurn: state.turnNumber, updatedAt: new Date() })
        .where(eq(gameState.roomId, params.roomId))

      return { success: true, state }
    },
    {
      params: t.Object({ roomId: t.String() }),
      body: t.Object({ cardIndex: t.Number() }),
    },
  )
